src/scroll-timeline-base.js

```diff
@@ -298,6 +313,9 @@ export class ScrollTimeline {
       // Internal members
       animations: [],
       scrollListener: null,
+
+      // Stored ranges
+      ranges: {}
```

Comment:
That might be a cleaner solution, indeed. 

We just need to store some more measurements about the subject, such as top and left. `calculateRange` currently iterates through all offset parents up to the scroll containers offsetParent, calling offsetLeft and offsetTop to calculate the relative position of the subject. This will invalidate the layout.

I'll look into it and update the PR

Reply:
It turns out that `calculateRange` is used in scroll-timeline-css.js to calculate keyframe offsets as well. It is used before any timelines have been instantiated, so we won't have access to any stored measurements here. 

https://github.com/flackr/scroll-timeline/blob/5a2639adcd5691004177b10bd9970f4cea4b4fc0/src/scroll-timeline-css.js#L96-L102

https://github.com/flackr/scroll-timeline/blob/5a2639adcd5691004177b10bd9970f4cea4b4fc0/src/scroll-timeline-css.js#L54-L58

We could refactor calculateRange to separate measurements and calculations, so that it can be used both places. 
The measurment and calculation of the subjects relative position should also be included in the stored measurements to prevent reflow:
https://github.com/flackr/scroll-timeline/blob/5a2639adcd5691004177b10bd9970f4cea4b4fc0/src/scroll-timeline-base.js#L522-L532

If we do this, the stored measurements will grow to include. 
```js
const measurements = {
  source: {
    scrollLeft,
    scrollTop,
    scrollWidth,
    scrollHeight,
    clientWidth,
    clientHeight,
    writingMode
  },
  subject: {
    top,
    left,
    offsetWidth,
    offsetHeight
  }
}
```

I'll start to look into it, hopefully next week.

Reply:
This sounds good to me. I was expecting we would include all required measurements in the structure so that once we've stored it we never access any layout inducing properties again.

I have a longer term goal to do this for all timelines before setting the progress in any animations so that we don't dirty the layout from one timeline update before the next one does more calculations but we don't have to do that in this patch.

Reply:
👍 

We are already storing source measurements and calculated ranges per source element (and not per timeline) before updating the timelines and ticking the animations. I don't think refactoring will change this, so we'll store both source measurements, subject measurements and relative positions, before we're updating the timelines and animations :)

Atm we're only observing changes to the source element. I'm not sure how we'll do this once we start observing the subject, but we'll look into this when we get there :)

Reply:
@flackr I have refactored the code now to separate measurement and range calculations.

On resize and style changed callbacks we first measure the source element any timeline subjects, before updating timelines and animations. Ranges are now calculated on the fly based on the measurements.

I have also switched to using weak refs for the set of timelines that are connected to a source element, so that timelines are not prevented from being garbage-collected.